internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go

```diff
@@ -46,19 +55,131 @@ func ContextProviderBuilder(logger *logger.Logger, c *config.Config, managed boo
 		return nil, errors.New(err, "failed to unpack configuration")
 	}
 	return &contextProviderK8sSecrets{
-		logger: logger,
-		config: &cfg,
+		logger:       logger,
+		config:       &cfg,
+		secretsCache: make(map[string]secretsData),
 	}, nil
 }
 
 func (p *contextProviderK8sSecrets) Fetch(key string) (string, bool) {
-	// key = "kubernetes_secrets.somenamespace.somesecret.value"
+	return p.getFromCache(key)
+}
+
+// Run initializes the k8s secrets context provider.
+func (p *contextProviderK8sSecrets) Run(ctx context.Context, comm corecomp.ContextProviderComm) error {
+	client, err := getK8sClientFunc(p.config.KubeConfig, p.config.KubeClientOptions)
+	if err != nil {
+		p.logger.Debugf("Kubernetes_secrets provider skipped, unable to connect: %s", err)
+		return nil
+	}
+	p.clientMx.Lock()
+	p.client = client
+	p.clientMx.Unlock()
+	go p.updateSecrets(ctx)
+
+	<-comm.Done()
+
+	p.clientMx.Lock()
+	p.client = nil
+	p.clientMx.Unlock()
+	return comm.Err()
+}
+
+func getK8sClient(kubeconfig string, opt kubernetes.KubeClientOptions) (k8sclient.Interface, error) {
+	return kubernetes.GetKubernetesClient(kubeconfig, opt)
+}
+
+// Update the secrets in the cache every TTL minutes
+func (p *contextProviderK8sSecrets) updateSecrets(ctx context.Context) {
+	d, err := time.ParseDuration(p.config.TTL)
+	if err != nil {
+		p.logger.Debugf(
+			"TTL %v is not valid: %v. Default of %v will be used.",
+			p.config.TTL,
+			err,
+			defaultTTL,
+		)
+		d, _ = time.ParseDuration(defaultTTL)
+	}
+	timer := time.NewTimer(d)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-timer.C:
+			p.updateCache()
+			timer.Reset(d)
+		}
+	}
+}
+
+func (p *contextProviderK8sSecrets) updateCache() {
+	p.secretsCacheMx.Lock()
+	for name, data := range p.secretsCache {
+		newValue, ok := p.fetchSecret(name)
+
+		// remove the secret from the cache
+		if !ok {
+			delete(p.secretsCache, name)
+		} else {
+			// if the secret has not been accessed in over 1h, delete it
+			// to avoid keeping secrets in cache that are no longer in use
+			diff := time.Since(data.lastAccess).Hours()
+			if diff > 1 {
```

Comment:
If you make the TTL configurable, why not make this configuration as a `time.Duration`?